lib/rubocop/cop/packaging/relative_require_to_lib.rb

```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+module RuboCop # :nodoc:
+  module Cop # :nodoc:
+    module Packaging # :nodoc:
+      # TODO
+      # Write about this cop.
+      # Write bad and good methods.
+      # Help from cop/lint/redundant_require_statement.rb.
+      #
+      class RelativeRequireToLib < Cop
+        # This is the message that will be displayed when RuboCop finds an
+        # offense of using `require` with relative path to lib.
+        MSG = 'Avoid using `require` with relative path to lib.'
+
+        def_node_matcher :require?, <<~PATTERN
+          (send nil? :require
+            (str {#starts_with_relative_path?}))
+        PATTERN
+
+        # Extended method.
+        # TODO: document it.
+        def on_send(node)
+          return unless require?(node)
+
+          add_offense(node)
+        end
+
+        # This method is called from inside `#def_node_search`.
+        # It is used to find the strings that uses relative path.
+        def starts_with_relative_path?(str)
+          str.match?(%r{.*\.\./lib/})
```

Comment:
With this regex, the following test (diff) is failing:
```
--- a/spec/rubocop/cop/packaging/relative_require_to_lib_spec.rb
+++ b/spec/rubocop/cop/packaging/relative_require_to_lib_spec.rb
@@ -14,6 +14,13 @@ RSpec.describe RuboCop::Cop::Packaging::RelativeRequireToLib do
     RUBY
   end
 
+  it 'registers an offense when using `require` with `File.expand_path`' do
+    expect_offense(<<~RUBY)
+      require File.expand_path('../lib/rubocop', __FILE__)
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ #{message}
+    RUBY
+  end
+
   it 'registers an offense when using `require` with relative path to lib/whatever/file' do
     expect_offense(<<~RUBY)
       require '../lib/whatever/file'
```

Reply:
:sob: :sob: :sob: 
No regex now :cactus: 

Different approach, altogether. Shall try to implement.

Reply:
Ah!
Using `File.expand_path` requires altogether a different AST, maybe something along the lines of:
```
def_node_matcher :require_file?, <<~PATTERN
  (send nil? :require
    (send
      (const {cbase nil?} :File) :expand_path
        (str {#starts_with_relative_path?})
        (str "(fragment:0)")))
PATTERN
```

Reply:
D'oh!
Using `__dir__` instead of __FILE__` makes the AST to look something like:
```
(send nil :require
  (send
    (const nil :File) :expand_path
    (str #starts_with_relative_path?)
    (send nil :__dir__)))
```

Reply:
Fixed via https://github.com/utkarsh2102/rubocop-packaging/pull/4/commits/fdbb59fd739f604b39b83f4b409341294b9263e9.